Intel tbb nvhpc issues

payerle opened the pull request:
fixes #25043

This PR adds two patches for use when building intel-tbb%nvhpc, as per #25043

One should remove compiler flags not recognized by nvhpc.  (I have not thoroughly tested as I have --noswitcherror set on my nvhpc compilers, but did not see any warnings in final spack-build-out.txt).  Also note that the  "-flifetime-dse=1" flag is supposed to work around a bug in intel-tbb (see https://bugs.archlinux.org/task/49898), so that bug might be present still (but as nvc++ has no equivalent option that I can find, I do not see any way around it).  (And the patch is only applied when using %nvhpc, so should not affect non %nvhpc builds)

The second patch (also only applied for %nvhpc) adds the --no_preincludes when nvc++ is being called as a preprocessor (-E option); this should prevent the pre-included headers from messing up the production of ld version scripts, etc.

payerle committed:
intel-tbb: Add patch when building with %nvhpc to remove unknown compiler flags

This should fix part of issue #25043

I confess, I use --noswitch error flag with my nvhpc compilers, so
I have not thoroughly tested this removes _all_ problematic flags,
although I did not see any warnings from such in my final spack-build-out.txt

payerle committed:
intel-tbb: Fix for %nvhpc to fix version script generation

When compiling with %nvhpc, add a patch to add the
--no_preincludes flag to PREPROC_ONLY make macro so that
the the generated version scripts, etc. do not get the
preincluded content, which was causing problems.

Should fix #25043

alalazo commented:
@fspiga @samcmill Do you have some cycles to review this?

samcmill commented on var/spack/repos/builtin/packages/intel-tbb/package.py:
```diff
@@ -120,6 +120,13 @@ class IntelTbb(Package):
 
+    # Support for building with %nvhpc
+    # 1) remove flags nvhpc compilers do not recognize
+    patch("intel-tbb.nvhpc-remove-flags.patch", when="%nvhpc")
```
Probably want to make both patches conditional on the tbb version.

samcmill reviewed (dismissed):
Other than the one comment, LGTM.

payerle committed:
intel-tbb: Added versioning to patches for nvhpcc

The nvhpc-remove-flags patches were rather version dependent

These are the versions tested (and successfully patched) against
intel-tbb.nvhpc-remove-flags.2017.patch: @2017, @2017.8, @2018, @2018.6
intel-tbb.nvhpc-remove-flags.2019.patch: @2019
intel-tbb.nvhpc-remove-flags.2019.1.patch: @2019.[1-6]
intel-tbb.nvhpc-remove-flags.2019.7.patch: @2019.[7-8]
intel-tbb.nvhpc-remove-flags.2019.9.patch: @2019.9, 2020.[0-3]

The intel-tbb.nvhpc-version-script-fix.2017.patch was tested and
applied successfully against all of the versions above.

payerle committed:
intel-tbb: Fix style issue

payerle committed:
intel-tbb: Undo some accidental changes

Due to test environment package.py being slightly out of date
with git master

alalazo commented:
@payerle Are there reports upstream for these issues when compiling with `nvhpc`?

payerle commented:
@alalazo Officially, intel-tbb only supports icc and gcc.  I have just submitted a ticket upstream describing what I did to compile it using nvhpc, _maybe_ they wil add nvhpc support???

adamjstewart commented on var/spack/repos/builtin/packages/intel-tbb/package.py:
```diff
@@ -10,7 +10,6 @@
 
 from spack import *
 
-
```
This change is causing flake8 tests to fail:
```
  var/spack/repos/builtin/packages/intel-tbb/package.py:13: [E302] expected 2 blank lines, found 1
```

payerle replied:
Should be fixed now

payerle committed:
intel-tbb: Fix another flake8 issue
